Sources/AsyncKit/ConnectionPool/EventLoopGroupConnectionPool.swift

```diff
@@ -178,14 +185,110 @@ public final class EventLoopGroupConnectionPool<Source> where Source: Connection
+    /// to deinitialize before it has fully shut down.
+    ///
+    /// This method promises explicitly as API contract not to invoke the callback before returning
+    /// to its caller. It further promises the callback will not be invoked on any event loop
+    /// belonging to the pool.
+    ///
+    /// - Warning: Any invocation of the callback represents a signal that the pool has fully shut
+    ///   down. This is true even if the error parameter is non-`nil`; errors are purely advisory.
+    public func shutdownGracefully(_ callback: @escaping (Error?) -> Void) {
+        // Protect access to shared state.
+        guard self.lock.withLock({
+            // Do not initiate shutdown multiple times.
+            guard !self.didShutdown else {
+                DispatchQueue.global().async {
+                    self.logger.warning("Connection pool can not be shut down more than once.")
+                    callback(ConnectionPoolError.shutdown)
+                }
+                return false
+            }
+            
+            // Set the flag as soon as we know a shutdown is needed.
+            self.didShutdown = true
+            self.logger.trace("Connection group pool shutdown start - telling the loop pools what's up.")
+            // Don't need to hold the lock anymore; the shutdown can proceed without blocking anything else, though
+            // it's also true there's nothing else to block that we care about after shutdown begin.
+            return true
+        }) else { return }
+        
+        // Tell each pool to shut down and take note of any errors if they show up. Use the dispatch
+        // queue to manage synchronization to avoid being trapped on any of our own event loops. When
+        // all pools are closed, invoke the callback and provide it the first encountered error, if
+        // any. By design, this loosely matches the general flow used by `MultiThreadedEventLoopGroup`'s
+        // `shutdownGracefully(queue:_:)` implementation.
+        let shutdownQueue = DispatchQueue(label: "codes.vapor.async-kit.poolShutdownGracefullyQueue")
+        let shutdownGroup = DispatchGroup()
+        var outcome: Result<Void, Error> = .success(())
+        
+        for pool in self.storage.values {
+            shutdownGroup.enter()
+            pool.close().whenComplete { result in
+                shutdownQueue.async() {
+                    outcome = outcome.flatMap { result }
+                    shutdownGroup.leave()
+                }
+            }
+        }
+        
+        shutdownGroup.notify(queue: shutdownQueue) {
+            switch outcome {
+                case .success:
+                    self.logger.debug("Connection group pool finished shutdown.")
+                    callback(nil)
+                case .failure(let error):
+                    self.logger.error("Connection group pool got shutdown error (and then shut down anyway): \(error)")
+                    callback(error)
```

Comment:
```suggestion
            case .success:
                self.logger.debug("Connection group pool finished shutdown.")
                callback(nil)
            case .failure(let error):
                self.logger.error("Connection group pool got shutdown error (and then shut down anyway): \(error)")
                callback(error)
```
Apparently it's easier to suggest things to myself than to make the changes any other way. Good grief, GitHub